presto-main/src/test/java/com/facebook/presto/cost/TestIntersectStatsRule.java

```diff
@@ -0,0 +1,252 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.cost;
+
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import org.testng.annotations.Test;
+
+import static com.facebook.presto.cost.SymbolStatsEstimate.ZERO_STATS;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+
+public class TestIntersectStatsRule
+        extends BaseStatsCalculatorTest
+{
+    @Test
+    public void testIntersectWhenRangesAreOverlapping()
+    {
+        tester().assertStatsFor(pb -> pb
+                .intersect(
+                        ImmutableList.<PlanNode>of(
+                                pb.values(pb.symbol("i11", BIGINT)),
+                                pb.values(pb.symbol("i21", BIGINT))),
+                        ImmutableListMultimap.<Symbol, Symbol>builder()
+                                .putAll(pb.symbol("o1", BIGINT), pb.symbol("i11", BIGINT), pb.symbol("i21", BIGINT))
+                                .build(),
+                        ImmutableList.of(pb.symbol("o1", BIGINT))))
+                .withSourceStats(0, PlanNodeStatsEstimate.builder()
+                        .setOutputRowCount(10)
+                        .addSymbolStatistics(new Symbol("i11"), SymbolStatsEstimate.builder()
+                                .setLowValue(0)
+                                .setHighValue(10)
+                                .setDistinctValuesCount(8)
+                                .setNullsFraction(0.2)
+                                .build())
+                        .build())
+                .withSourceStats(1, PlanNodeStatsEstimate.builder()
+                        .setOutputRowCount(20)
+                        .addSymbolStatistics(new Symbol("i21"), SymbolStatsEstimate.builder()
+                                .setLowValue(5)
+                                .setHighValue(15)
+                                .setNullsFraction(0.4)
+                                .setDistinctValuesCount(4)
+                                .build())
+                        .build())
+                .check(check -> check
+                        .outputRowsCount(5)
+                        .symbolStats("o1", assertion -> assertion
+                                .lowValue(5)
+                                .highValue(10)
+                                .dataSizeUnknown()
+                                .distinctValuesCount(4)
+                                .nullsFraction(0.2)));
+    }
+
+    @Test
+    public void testIntersectWhenRangesAreSeparated()
```

Comment:
We compute 1 instead of 0 because 1 is the minimal row count result of `AggregationStatsRule#groupBy`.
We could fix this.. but what if we calculate something incorrectly and 0 row count turns into some damage later on?
just thinking..

@kokosing do you want to add something?

Reply:
Why have an IntersectStatsRule at all then?  Maybe we should get rid of it if it's not going to be used.

Reply:
I do not remember why it was created. Maybe we was not aware of this rewrite. Or maybe we had too much processing power in Teradata ;)

From the other point of view. If we have then what is the benefit of its removal? I could image a CBO rule that could leverage this at some point in the future. For example, I believe there are multiple ways of how Intersect could be implemented and deciding which way should be used could be a cost-based decision.

Reply:
I agree, if this is not needed now, we should remove it. We were generally cautious not to do things within CBO that are already handled outside.